File: sobit_light_library/src/sobit_light_wheel_controller.cpp
#include "sobit_light_library/sobit_light_wheel_controller.hpp"


namespace sobit_light {

WheelController::WheelController ( const std::string &name )
: Node( name )
{
    if (!this->get_node_base_interface()) {
        throw std::runtime_error("Failed to initialize ROS2 Node.");
    }
     rclcpp::executors::SingleThreadedExecutor executor;
    executor.add_node(this->shared_from_this());
    executor.spin_some();
    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>(
        "/odom", 1, std::bind(&WheelController::callbackOdometry, this, std::placeholders::_1));
    pub_cmd_vel_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel_mux/input/teleop", 1);

    rclcpp::spin_some(this->get_node_base_interface());
    rclcpp::sleep_for(std::chrono::seconds(3));
}

WheelController::WheelController ()
: Node("sobit_light_wheel_controller")
{
    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>(
        "/odom", 1, std::bind(&WheelController::callbackOdometry, this, std::placeholders::_1));
    pub_cmd_vel_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel_mux/input/teleop", 1);

    rclcpp::spin_some(this->get_node_base_interface());
    rclcpp::sleep_for(std::chrono::seconds(3));
}

WheelController::WheelController(const rclcpp::NodeOptions & options)
: Node("sobit_light_wheel_controller", options) {
    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>(
        "/odom", 1, std::bind(&WheelController::callbackOdometry, this, std::placeholders::_1));
    pub_cmd_vel_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel_mux/input/teleop", 1);
    rclcpp::spin_some(this->get_node_base_interface());
    rclcpp::sleep_for(std::chrono::seconds(3));
}

bool WheelController::controlWheelLinear( const double distance ) {
    try {
        auto start_time = this->get_clock()->now();
        geometry_msgs::msg::Twist output_vel;

        while ( curt_odom_.pose.pose.position.x == 0 &&
                curt_odom_.pose.pose.position.y == 0 &&
                curt_odom_.pose.pose.position.z == 0 &&
                curt_odom_.pose.pose.orientation.w == 0 ) {

            rclcpp::spin_some(this->get_node_base_interface());
        }
        nav_msgs::msg::Odometry init_odom = curt_odom_;

        double moving_distance = 0.0;
        double target_distance = std::fabs( distance );

        double Kp = 0.1;
        double Ki = 0.4;
        double Kd = 0.8;

        double velocity_differential = Kp * distance;
        rclcpp::Rate loop_rate(20);

        while ( moving_distance < target_distance  ) {
            rclcpp::spin_some(this->get_node_base_interface());

            auto end_time = this->get_clock()->now();
            auto elapsed_time = (end_time - start_time).seconds();

            double vel_linear = 0.0;

            // PID
            if ( target_distance <= 0.1 ) {
                vel_linear =  Kp * ( target_distance + 0.001 - moving_distance ) - Kd * velocity_differential + Ki / 0.8 * ( target_distance + 0.001 - moving_distance ) * std::pow( elapsed_time, 2 );
            } else {
                vel_linear =  Kp * ( target_distance + 0.001 - moving_distance ) - Kd * velocity_differential + Ki / ( 8.0 / target_distance ) * ( target_distance + 0.001 - moving_distance ) * std::pow( elapsed_time, 2 );
            } 
            output_vel.linear.x = ( distance > 0 ) ? vel_linear : -vel_linear;
            velocity_differential = vel_linear;
            
            pub_cmd_vel_->publish( output_vel );

            double x_diif = curt_odom_.pose.pose.position.x - init_odom.pose.pose.position.x;
            double y_diif = curt_odom_.pose.pose.position.y - init_odom.pose.pose.position.y;
            moving_distance = std::hypot( x_diif, y_diif );

            RCLCPP_INFO(this->get_logger(), "x_diif = %f\tx2_diif = %f\ty_diif = %f\ty2_diif = %f", init_odom.pose.pose.position.x, curt_odom_.pose.pose.position.x, init_odom.pose.pose.position.y, curt_odom_.pose.pose.position.y );
            loop_rate.sleep();
        }

        return true;
    } catch ( const std::exception& ex ) {
        RCLCPP_ERROR(this->get_logger(), "%s", ex.what());
        return false;
    }
}

bool WheelController::controlWheelRotateRad( const double angle_rad ) {
    try {
        while ( curt_odom_.pose.pose.position.x == 0 &&
                curt_odom_.pose.pose.position.y == 0 &&
                curt_odom_.pose.pose.position.z == 0 &&
                curt_odom_.pose.pose.orientation.w == 0 ) {

            rclcpp::spin_some(this->get_node_base_interface());
        }

        auto start_time = this->get_clock()->now();

        geometry_msgs::msg::Twist output_vel;
        double init_yaw = geometryQuat2Yaw( curt_odom_.pose.pose.orientation );
        double moving_angle_rad = 0.0;
        double abs_angle_rad =  std::fabs( angle_rad );
        double abs_angle_deg = rad2Deg( abs_angle_rad );

        double Kp = 0.1;
        double Ki = 0.4;
        double Kd = 0.8;

        double velocity_differential = Kp * angle_rad;
        int loop_cnt = 1;
        rclcpp::Rate loop_rate(20);

        while ( moving_angle_rad < abs_angle_rad ) {
            rclcpp::spin_some(this->get_node_base_interface());

            auto end_time = this->get_clock()->now();
            auto elapsed_time = (end_time - start_time).seconds();
            double vel_angular = 0.0;

            // PID
            if ( abs_angle_deg <= 30 ) {
                vel_angular = Kp * ( abs_angle_rad + 0.001 - moving_angle_rad )
                            - Kd * velocity_differential
                            + Ki * ( abs_angle_rad + 0.001 - moving_angle_rad ) * std::pow( elapsed_time, 2 );
            } else {
                vel_angular = Kp * ( abs_angle_rad + 0.001 - moving_angle_rad )
                            - Kd * velocity_differential
                            + Ki * ( abs_angle_rad + 0.001 - moving_angle_rad ) * std::pow( elapsed_time, 2 ) * 0.75 * 30 / abs_angle_deg;
            }
            output_vel.angular.z = ( angle_rad > 0 ) ? vel_angular : - vel_angular;
            velocity_differential = vel_angular;
            
            pub_cmd_vel_->publish( output_vel );
            
            double curt_yaw = geometryQuat2Yaw( curt_odom_.pose.pose.orientation );
            double pre_move_ang_rad = moving_angle_rad;

            if ( -0.00314 < curt_yaw - init_yaw && curt_yaw - init_yaw < 0 && 0 < angle_rad ) continue;
            else if ( 0 < curt_yaw - init_yaw && curt_yaw - init_yaw < 0.00314 && angle_rad < 0 ) continue;

            if ( curt_yaw - init_yaw < 0 && 0 < angle_rad ) moving_angle_rad = std::abs(curt_yaw - init_yaw + deg2Rad(360 * loop_cnt));
            else if ( 0 < curt_yaw - init_yaw && angle_rad < 0 ) moving_angle_rad = std::abs(curt_yaw - init_yaw - deg2Rad(360 * loop_cnt));
            else if ( 0 < angle_rad ) moving_angle_rad = std::abs(curt_yaw - init_yaw + deg2Rad(360 * (loop_cnt-1)));
            else moving_angle_rad = std::abs(curt_yaw - init_yaw - deg2Rad(360 * (loop_cnt-1)));

            if ( rad2Deg(moving_angle_rad) < (rad2Deg(pre_move_ang_rad)-0.0314) ) {
                loop_cnt++;
                if ( 0 < angle_rad ) moving_angle_rad = std::abs(curt_yaw - init_yaw + deg2Rad(360 * (loop_cnt-1)));
                else moving_angle_rad = std::abs(curt_yaw - init_yaw - deg2Rad(360 * (loop_cnt-1)));
            }
            
            RCLCPP_INFO(this->get_logger(), "target_angle = %f\tmoving_angle = %f", abs_angle_rad, moving_angle_rad );
            
            loop_rate.sleep();
        }
        return true;
    } catch ( const std::exception& ex ) {
        RCLCPP_ERROR( this->get_logger(), "%s", ex.what() );
        return false;
    }
}

bool WheelController::controlWheelRotateDeg( const double angle_deg ) {
    return controlWheelRotateRad( deg2Rad(angle_deg) );
}

}

RCLCPP_COMPONENTS_REGISTER_NODE(sobit_light::WheelController)

File: sobit_light_library_python/sobit_light_library_python/sobit_light_wheel_controller.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from math import (radians, degrees, fmod, copysign, pi)

import rclpy
from rclpy.node import Node

from tf_transformations import euler_from_quaternion

from nav_msgs.msg import Odometry
from geometry_msgs.msg import Twist

def geoQuat2Yaw(geo_quat):
  quat_list = [geo_quat.x, geo_quat.y, geo_quat.z, geo_quat.w]

  (roll, pitch, yaw) = euler_from_quaternion(quat_list)

  return yaw


class WheelController(Node):
  def __init__(self, node_name='sobit_light_wheel_controller'):
    super().__init__(node_name)

    # QoS profile for the subscription
    qos_policy = rclpy.qos.QoSProfile(
        reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT,
        history=rclpy.qos.HistoryPolicy.KEEP_LAST,
        depth=1)
    
    # 
    self.sub_odom_ = self.create_subscription(
        Odometry,
        '/kachaka/odometry/odometry',
        self.callbackOdometry,
        qos_profile=qos_policy,
    )

    self.pub_cmd_vel_ = self.create_publisher(
        Twist,
        '/kachaka/manual_control/cmd_vel',
        qos_profile=qos_policy,
    )

    # Flags
    self.is_running_ = False

    # Variables
    self.curt_odom_ = Odometry()

    # Rate
    self._loop_rate = self.create_rate(20, self.get_clock())

  def callbackOdometry(self, msg):
    self.is_running_ = True
    self.curt_odom_ = msg

  def controlWheelLinear(self, distance):
    while not(self.is_running_):
      if not rclpy.ok():
        self.sub_odom_.destroy()
      self.get_logger().info('Waiting for odometry...')
      rclpy.spin_once(self)

    # Get the initial position
    init_odom = self.curt_odom_
    self.get_logger().info('Initial odometry: %s' % init_odom)

    # Get the current time
    start_time = self.get_clock().now()

    # Initialize values
    output_vel = Twist()
    moved_distance = 0.0
    target_distance = abs(distance)

    # PID Values
    Kp = 0.1
    Ki = 0.4
    Kd = 0.8

    vel_diff = Kp * distance

    while (moved_distance < target_distance):
      # Get the current time
      curr_time = self.get_clock().now()

      # Calculate the elapsed time
      elapsed_time = curr_time - start_time
      elapsed_time = elapsed_time.nanoseconds
      elapsed_time = elapsed_time / 1e9

      vel_linear = 0.0

      # TODO: Fix the PID controller
      if (target_distance < 0.01) or (target_distance - moved_distance < 0.01):
        break
      if (target_distance < 0.69) and (abs(vel_diff) > 1.0):
        vel_diff = copysign(1.0, vel_diff)

      if (target_distance < 0.1):
        vel_linear = Kp * (target_distance + 0.001 - moved_distance) \
                   - Kd * vel_diff \
                   + Ki / 0.8 * (target_distance + 0.001 - moved_distance) * pow(elapsed_time, 2)
      else:
        vel_linear = Kp * (target_distance + 0.001 - moved_distance) \
                   - Kd * vel_diff \
                   + Ki / (8.0 / target_distance) * (target_distance + 0.001 - moved_distance) * pow(elapsed_time, 2)

      output_vel.linear.x = vel_linear if distance > 0 else -vel_linear
      vel_diff = vel_linear

      self.pub_cmd_vel_.publish(output_vel)

      # Calculate the moved distance
      x_diff = self.curt_odom_.pose.pose.position.x - init_odom.pose.pose.position.x
      y_diff = self.curt_odom_.pose.pose.position.y - init_odom.pose.pose.position.y
      moved_distance = pow(pow(x_diff, 2) + pow(y_diff, 2), 0.5)

      # Debug
      self.get_logger().info('[Wheel Control: Linear] Moved distance: %f, Target distance: %f' % (moved_distance, target_distance))
      rclpy.spin_once(self)
    
    # Stop the robot after the target distance is reached
    output_vel.linear.x = 0.0
    self.pub_cmd_vel_.publish(output_vel)

  def controlWheelRotateRad(self, angle_rad):
    while not(self.is_running_):
      if not rclpy.ok():
        self.sub_odom_.destroy()
      self.get_logger().info('Waiting for odometry...')
      rclpy.spin_once(self)

    # Get the initial position
    init_odom = self.curt_odom_
    init_yaw = geoQuat2Yaw(init_odom.pose.pose.orientation)
    self.get_logger().info('Initial odometry: %s' % init_odom)

    # Get the current time
    start_time = self.get_clock().now()

    # Initialize values
    output_vel = Twist()
    moved_angle_rad = 0.0
    target_angle_rad = abs(angle_rad)
    target_angle_deg = degrees(target_angle_rad)

    # PID Values
    Kp = 0.1
    Ki = 0.4
    Kd = 0.8

    vel_diff = Kp * angle_rad

    while (moved_angle_rad < target_angle_rad-0.01):
      rclpy.spin_once(self)

      # Get the current time
      curr_time = self.get_clock().now()

      # Calculate the elapsed time
      elapsed_time = curr_time - start_time
      elapsed_time = elapsed_time.nanoseconds
      elapsed_time = elapsed_time / 1e9
      # print(elapsed_time)

      vel_angular = 0.0

      if (target_angle_deg < 30):
        vel_angular = Kp * (target_angle_rad + 0.001 - moved_angle_rad) \
                   - Kd * vel_diff \
                   + Ki / 0.8 * (target_angle_rad + 0.001 - moved_angle_rad) * pow(elapsed_time, 2)
      else:
        vel_angular = Kp * (target_angle_rad + 0.001 - moved_angle_rad) \
                   - Kd * vel_diff \
                   + Ki / (8.0 / target_angle_rad) * (target_angle_rad + 0.001 - moved_angle_rad) * pow(elapsed_time, 2)

      output_vel.angular.z = vel_angular if angle_rad > 0 else -vel_angular
      vel_diff = vel_angular
      self.pub_cmd_vel_.publish(output_vel)

      # Calculate the moved distance
      curt_yaw = geoQuat2Yaw(self.curt_odom_.pose.pose.orientation)
      self.get_logger().info('Current yaw: %f, Initial yaw: %f' % (curt_yaw, init_yaw))

      angle_diff_rad = curt_yaw - init_yaw
      angle_diff_rad = fmod(angle_diff_rad, 2*pi)
      moved_angle_rad = abs(angle_diff_rad)

      # Debug
      self.get_logger().info('[Wheel Control: Rotate] Moved angle: %f, Target angle: %f' % (moved_angle_rad, target_angle_rad))

      # Sleep
      # _loop_rate.sleep()

    # Stop the robot after the target angle is reached
    output_vel.angular.z = 0.0
    self.pub_cmd_vel_.publish(output_vel)

  def controlWheelRotateDeg(self, angle_deg):
    self.controlWheelRotateRad(radians(angle_deg))


def main(args=None):
  rclpy.init(args=args)

  node = WheelController()

  node.controlWheelLinear(distance=0.5)
  node.controlWheelRotateRad(angle_rad=1.57)
  node.controlWheelRotateDeg(angle_deg=-90)
  node.controlWheelLinear(distance=-0.5)

  node.destroy_node()
  rclpy.shutdown()

if __name__ == '__main__':
    main()
